### DISCUSSION TITLE
Status & Todo & Progress

### DISCUSSION THREAD

**Discussion by ZhiyuanLck** [ZhiyuanLck/smart-pairs#17] [Ideas]
## Current Status

Busy with work.

## Plan

- ~Refactor deadline: 7.1~

## Progress

- [ ] Refactor with `C` multi-threading support
  - [ ] Preparation work in `lua`
    - [x] Configuration
    - [ ] New config frame work
    - [x] Check if the current cursor is in skip regions
    - [ ] Pass lines and correct pairs to `C` parsers
  - [x] Base data structure to store pairs
  - [ ] Design new algorithm to count the pairs
    - [x] Record the position of the pair and store it in a linked list
    - [x] Parse the pairs with the list
    - [ ] Special cases 1: inside strings
    - [ ] Special cases 2: inside comments
    - [ ] Interact with `lua`
    - [ ] Benchmark
  - [ ] Multi-threading support
    - [x] Thread pool
    - [x] ~~Mutex of the pair list~~ lock-free
    - [ ] Benchmark
- [ ] Features
  - [ ] Integrate [`endwise`](https://github.com/tpope/vim-endwise) #9
  - [ ] Take control over normal enter #13
  - [ ] Setup with completion plugin #2
  - [ ] Smart equal #26
  - [ ] Disable specific file types in global pairs #31
  - [ ] Smart jump #23
  - [ ] Align right bracket with cross-line left bracket.
  - [ ] Text object
  - [ ] Better `matchpairs`
  - [ ] Pair hint
- [ ] Document
  - [ ] Add more doc for advanced usage
  - [ ] Add gifs

Welcom to comment with any fancy ideas!

**Comment by bryant-the-coder** [comment]
space out the equal sign

**Comment by ignamartinoli** [comment]
We could add custom pairs for **HTML** comments and **Haskell** annotations
EDIT: also thanks for the plugin, is the most feature rich pair plugin right now
